storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -1733,13 +1825,24 @@ private TopologyResources getResourcesForTopology(String topoId, StormBase base)
         return ret;
     }
 
+    private boolean isReadyForMKAssignments() throws Exception {
+        if(isLeader()) {
+            if(isHeartbeatsRecovered()) {
+                return true;
+            }
+            LOG.info("waiting for worker heartbeats recovery, skipping assignments");
```

Comment:
We could change this to `warn`, cause Nimbus leader should be always ready to do assignments. No assignment due to leader is normal and expected.

Reply:
done